src/foam/nanos/notification/email/SMTPEmailService.js

```diff
@@ -243,160 +200,49 @@ return config_;`
           type: 'foam.nanos.notification.email.EmailMessage'
         }
       ],
-      javaCode: `
-if ( ! this.getEnabled() ) return;
-
-((FixedThreadPool) getThreadPool()).submit(x, new ContextAgent() {
-  @Override
-  public void execute(X x) {
-    try {
-      MimeMessage message = createMimeMessage(finalizeEmailConfig(x, emailMessage));
-      if ( message == null ) {
-        return;
-      }
-
-      // send message
-      Transport transport = session_.getTransport("smtp");
-      transport.connect();
-      transport.sendMessage(message, message.getAllRecipients());
-      transport.close();
-    } catch (Throwable t) {
-      t.printStackTrace();
-    }
-  }
-});`
-    },
-    {
-      name: 'sendEmailFromTemplate',
-      javaCode: `
-if ( ! this.getEnabled() ) return;
-
-String group = user != null ? (String) user.getGroup() : null;
-EmailTemplate emailTemplate = DAOResourceLoader.findTemplate(getX(), name, group);
-if ( emailMessage == null )
-  return;
-
-for ( String key : templateArgs.keySet() ) {
-  Object value = templateArgs.get(key);
-  if ( value instanceof String ) {
-    String s = (String) value;
-    templateArgs.put(key, new String(s.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8));
-  }
-}
-
-EnvironmentConfiguration config = getConfig(group);
-JtwigModel model = JtwigModel.newModel(templateArgs);
-emailMessage = (EmailMessage) emailMessage.fclone();
-
-JtwigTemplate templateBody =    JtwigTemplate.inlineTemplate(emailTemplate.getBody(), config);
-emailMessage.setBody(templateBody.render(model));
-
-// If subject has already provided, then we don't want to use template subject.
-if (SafetyUtil.isEmpty(emailMessage.getSubject())) {
-  JtwigTemplate templateSubject = JtwigTemplate.inlineTemplate(emailTemplate.getSubject(), config);
-  emailMessage.setSubject(templateSubject.render(model));
-}
+      javaCode:
+        `
+        if ( ! this.getEnabled() ) return;
+        if ( emailMessage.getTo().length < 1 || SafetyUtil.isEmpty(emailMessage.getFrom())) {
+          throw new IllegalStateException("Email that is being transported out does not have basic fields set.");
+        }
 
-sendEmail(x, emailMessage);`
+        ((FixedThreadPool) getThreadPool()).submit(x, new ContextAgent() {
```

Comment:
Oh this kicks it out into its own thread huh? I'm not sure I like that happening in here. When I call `sendEmail` on the SMTPEmailService, I'd expect the email to be sent when the method returns. I think you'll want to make a decorator that calls an EmailService in a separate thread so we can move this feature around in the chain of EmailServices we're creating.

Reply:
![image](https://user-images.githubusercontent.com/10802216/56672887-31945900-6685-11e9-82a3-520e3bcce4c3.png)

So don't really get the point of a decorator. Why don't we just remove the thread?

Reply:
I'm guessing it was put in because SMTP needs to hit the network to actually send the email and we don't want it blocking. SMTPEmailService shouldn't care about this.